### PR TITLE
i880 - database work for the annex

### PR DESCRIPTION
**Pull request by richfitz**
This is the approach I am thinking we can use for splitting the database.  It's not ready to use yet because the schema migrations aren't there.  However, I think it's ready for a first round of review so that then I can move on to getting the provisioning done and because the schema changes will also happen here

I have gone with the name "annex" here to distinguish this from the main database (following from "git annex").  I hope this name will make it easier to talk about things!

Username and password to get at the annex are handled by the flyway configuration file and should be overridden on the cli via arguments like `-placeholders.montagu_db_annex_host=db-annex`.  The default are set up to be reasonably easy to set up testing with.

There is a stripped down docker compose configuration that can be used for testing in `annex/` - see the readme and scripts there.

## Lessons learnt and other things

* `COPY` will not work with foreign tables, and some tools will not recognise them as real tables (which they aren't really).  This may cause problem with the SQL drivers the API uses and will be worth testing
* The fastest way of getting data into the table is a `COPY` but using a second database connection.  This is unfortunate but I think unavoidable if we want reasonable speed.  The alternative is a `COPY` into a temporary table the main database, then `INSERT INTO` from that table
* Foreign keys are _hard_ and I'm glad we don't need them.  The commit ecab752 shows one approach that might have worked to get foreign keys here but it's a huge faff and requires syncing tables from them main database into the annex.  This was kind of fun to learn and might come in useful one day.

## Pending:

* Need to make the foreign table read only for _all_ users
* Schema migrations still to come - this just duplicates the existing structure as a reasonable starting point
* Add the testing bits as a patch to our main docker compose file which will help with testing (the idea being we'd spin up an extra container for the annex to entirely replicate how the data moves around in testing - this should be easy to do with an extra `-f` argument to docker compose).

### CODE REVIEW

File: annex/annex.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+function cleanup {
+    set +e
+    docker-compose stop
+    docker-compose rm -f -v
+    docker network rm $MONTAGU_NETWORK
+    docker volume rm "${MONTAGU_BASENAME}_db_volume" "${MONTAGU_BASENAME}_db_annex_volume"
+trap cleanup EXIT
+
+MONTAGU_DB_VERSION=i880 docker-compose up -d
```
Comment (MartinEden) [inline annex/annex.sh]: Is this just missing a new line?

File: bin/annex_sync [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#
+# at the end of the CREATE SUBSCRIPTION block here.  We can also
+# change this later with ALTER SUBSCRIPTION.
+#
+# https://www.postgresql.org/docs/10/static/sql-createsubscription.html
+# https://www.postgresql.org/docs/10/static/runtime-config-wal.html#guc-synchronous-commit
+def subscribe_start(d):
+    # This can't be done in a transaction, which given what it does is
+    # far from ideal!  If anything fails through this *will* leave the
+    # db in an unfortunately state.
```
Comment (MartinEden) [inline bin/annex_sync]: This feels like something we need to come up with a fix for.

Comment (richfitz) [reply bin/annex_sync]: Ah, this _whole file_ should not be here - it should have been purged in ecab752

File: annex/annex.sh (hunk 1 above)
Comment (richfitz) [reply annex/annex.sh]: I don't think so - what looks missing to you?

File: bin/annex_sync (hunk 2 above)
Comment (richfitz) [reply bin/annex_sync]: removed now

File: annex/annex.sh (hunk 1 above)
Comment (MartinEden) [reply annex/annex.sh]: So are you assigning the text `i880 docker-compose up -d` to `MONTAGU_DB_VERSION`? Are you assigning the variable and then running docker-compose? What's going on here?

Comment (richfitz) [reply annex/annex.sh]: This sets the variable `MONTAGU_DB_VERSION` to the value `i880` for the next command; the docker compose file references it (following the main montagu docker compose file)

**Comment by littledice** [comment]
Looks good; I haven't had a chance to run this locally and see if I can get it working but will do so on Monday.

**Comment by richfitz** [comment]
This is ready for review (and potentially for merge).  This set of changes provides mostly groundwork for the main montagu deploy work (separate PR on the montagu repo).

* Files in annex are now mostly concerned with setting up and deploying on the annex machine.  This is currently barebones and needs work when we have something we need to backup, but I would like to defer that work
* the code in annex/testing *probably* still works but I have not looked (but see the montagu repo for things working more generally)
* User management: we have a `vimc` user who can do whatever and a `readonly` user designed to be used from the main database

**Comment by richfitz** [comment]
One thing that does need doing before merge - update the filename of the sql migrations to match the correct date

File: annex/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    echo "montagu db annex volume already exists"
+    INITIAL_DEPLOY=0
+else
+    echo "Creating montagu db annex volume"
+    INITIAL_DEPLOY=1
+    docker volume create $ANNEX_VOLUME_NAME
+fi
+
+docker pull $ANNEX_IMAGE
+docker pull $ANNEX_MIGRATE_IMAGE
```
Comment (richfitz) [inline annex/deploy.sh]: this can come out

**Pull request merged**